### PR TITLE
Squash supermajority root on blockstore replay at startup

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
After long periods of no progress, a single node can have a long ledger it needs to replay, which is prohibitively taxing on memory and compute

#### Summary of Changes
Squash banks that supermajority has rooted on replay at startup

Fixes #

### CODE REVIEW

File: ledger/src/blockstore_processor.rs [hunk 1]
```diff
@@ -1535,11 +1627,11 @@ pub mod tests {
         let blockstore =
             Blockstore::open(&ledger_path).expect("Expected to successfully open database ledger");
 
-        // Let last_slot be the number of slots in the first two epochs
+        // Let `last_slot` be the number of slots in the first two epochs
         let epoch_schedule = get_epoch_schedule(&genesis_config, Vec::new());
         let last_slot = epoch_schedule.get_last_slot_in_epoch(1);
 
-        // Create a single chain of slots with all indexes in the range [0, last_slot + 1]
+        // Create a single chain of slots with all indexes in the range [0, v+ 1]
```
Comment (mvines) [inline ledger/src/blockstore_processor.rs:1634]: nit:
```suggestion
        // Create a single chain of slots with all indexes in the range [0, v + 1]
```

File: ledger/src/blockstore_processor.rs [hunk 2]
```diff
@@ -844,6 +889,51 @@ fn load_frozen_forks(
...
+    }
+
+    None
+}
+
+fn supermajority_root_from_bank(bank: &Bank) -> Option<Slot> {
+    let roots: BTreeMap<Slot, u64> = bank
+        .vote_accounts()
+        .into_iter()
+        .filter_map(|(key, (lamports, account))| {
```
Comment (mvines) [inline ledger/src/blockstore_processor.rs:909]:
```suggestion
        .filter_map(|(key, (stake, account))| {
```
I misread `lamports` at first to mean the number of lamports in that vote account, which doesn't work below at all!  Naming the variable as I suggest would have avoided my confusion

**Comment by sakridge** [comment]
Is it the best way to do this? Seems like a lot to keep in sync with how roots are actually set on the replay side.

**Comment by carllin** [comment]
@sakridge yeah I think it's ok because it's equivalent to restarting at a snapshot with a higher root than your local root.

With @ryoqun's persistent tower fixes, such scenarios will become safer for the cluster across restarts

File: ledger/src/blockstore_processor.rs [hunk 3]
```diff
@@ -844,6 +889,51 @@ fn load_frozen_forks(
...
+                    key,
+                    bank.slot()
+                );
+                return None;
+            }
+
+            vote_state
+                .unwrap()
+                .root_slot
+                .map(|root_slot| (root_slot, stake))
```
Comment (ryoqun) [inline ledger/src/blockstore_processor.rs:927]: @carllin Well is this ok? What if different vote account has same root_slot? I think it's silently overwritten when `collect()`-ing into `roots`? Or does `collect()` summing automatically for us in this case?

File: ledger/src/blockstore_processor.rs [hunk 4]
```diff
@@ -810,14 +820,49 @@ fn load_frozen_forks(
...
+                })
+            } else if blockstore.is_root(slot) {
+                Some(&bank)
+            } else {
+                None
+            }
+        };
+
+        if let Some(new_root_bank) = new_root_bank {
+            *root = new_root_bank.slot();
```
Comment (ryoqun) [inline ledger/src/blockstore_processor.rs:856]: Btw, how the non-empty slot interval `(blockstore.max_root(), supermajority_root]` is marked at rooted for the blockstore?

Comment (carllin) [reply ledger/src/blockstore_processor.rs:856]: @ryoqun it's currently not. We could, but I think the current implementation is equivalent to starting a snapshot at `supermajority_root`, in which case those rooted slots in the range `(blockstore.max_root(), supermajority_root]` won't be marked in blockstore anyways.

File: ledger/src/blockstore_processor.rs (hunk 3 above)
Comment (carllin) [reply ledger/src/blockstore_processor.rs:927]: @ryoqun that's a bug, it definitely doesn't sum it automatically. Thanks for the catch!

File: ledger/src/blockstore_processor.rs [hunk 5]
```diff
@@ -810,14 +820,49 @@ fn load_frozen_forks(
...
-            last_root_slot = slot;
+        // Block must be frozen by this point, otherwise `process_single_slot` would
+        // have errored above
+        assert!(bank.is_frozen());
+        all_banks.insert(bank.slot(), bank.clone());
+
+        // If we've reached the last known root in blockstore, start looking
+        // for newer cluster confirmed roots
+        let new_root_bank = {
+            if *root == max_root {
```
Comment (ryoqun) [inline ledger/src/blockstore_processor.rs:831]: (for the record:) this condition was wrong... #14557

**Pull request merged**